api/server/runner_httptrigger.go

```diff
@@ -112,15 +112,23 @@ func (trw *triggerResponseWriter) WriteHeader(statusCode int) {
 		realHeaders[k] = vs
 	}
 
-	// XXX(reed): simplify / add tests for these behaviors...
-	gatewayStatus := 200
-	if statusCode >= 400 {
-		gatewayStatus = 502
-	} else if fnStatus > 0 {
-		gatewayStatus = fnStatus
+	trw.inner.WriteHeader(determineFinalStatus(statusCode, fnStatus))
+}
+
+func determineFinalStatus(serviceStatus, userFnStatus int) int {
+	finalStatus := 200
+
+	if userFnStatus > 0 {
+		if userFnStatus >= 500 {
+			finalStatus = 502
+		} else {
+			finalStatus = userFnStatus
+		}
+	} else if serviceStatus > 0 {
+		finalStatus = serviceStatus
```

Comment:
I agree with @zootalures regarding precedence here. I think the logic here simply should be:

```
finalCode := serviceCode
if serviceCode < 400 {
    // copy over Fn-Http-Status or 502
}
```

Which I think doesn't change the outcome, but it's better to be defensive and guard against LB&Runner that sets a service error despite the existence of Fn-Http-Status header.